compositor_render/src/renderer/common_pipeline/surface.rs

```diff
@@ -0,0 +1,226 @@
+use wgpu::{util::DeviceExt, Buffer, BufferSlice};
+
+use super::{Vertex, MAX_TEXTURES_COUNT};
+
+/// In case of no input texture, vertex shader receives plane
+/// with 4 vertices with input id -1. This allows using shaders without
+/// any input textures - e.g. shaders generating some texture
+/// based on uniform parameters.
+const SINGLE_PLANE_VERTICES: [Vertex; 4] = [
+    Vertex {
+        position: [1.0, -1.0, 0.0],
+        texture_coords: [1.0, 1.0],
+        input_id: -1,
+    },
+    Vertex {
+        position: [1.0, 1.0, 0.0],
+        texture_coords: [1.0, 0.0],
+        input_id: -1,
+    },
+    Vertex {
+        position: [-1.0, 1.0, 0.0],
+        texture_coords: [0.0, 0.0],
+        input_id: -1,
+    },
+    Vertex {
+        position: [-1.0, -1.0, 0.0],
+        texture_coords: [0.0, 1.0],
+        input_id: -1,
+    },
+];
+
+#[rustfmt::skip]
+const SINGLE_PLANE_INDICES: [u16; 6] = [
+    0, 1, 2,
+    2, 3, 0
+];
+
+pub const INDEX_FORMAT: wgpu::IndexFormat = wgpu::IndexFormat::Uint16;
+
+struct Vertices<const N: usize>([Vertex; N]);
+
+impl<const N: usize> Vertices<N> {
+    const fn new() -> Self {
+        let mut vertices = [Vertex::empty(); N];
+
+        let mut input_id = 0;
+        while input_id < N / 4 {
+            vertices[input_id * 4] = Vertex {
+                position: [1.0, -1.0, 0.0],
+                texture_coords: [1.0, 1.0],
+                input_id: input_id as i32,
+            };
+
+            vertices[input_id * 4 + 1] = Vertex {
+                position: [1.0, 1.0, 0.0],
+                texture_coords: [1.0, 0.0],
+                input_id: input_id as i32,
+            };
+
+            vertices[input_id * 4 + 2] = Vertex {
+                position: [-1.0, 1.0, 0.0],
+                texture_coords: [0.0, 0.0],
+                input_id: input_id as i32,
+            };
+
+            vertices[input_id * 4 + 3] = Vertex {
+                position: [-1.0, -1.0, 0.0],
+                texture_coords: [0.0, 1.0],
+                input_id: input_id as i32,
+            };
+
+            input_id += 1;
+        }
+
+        Self(vertices)
+    }
+}
```

Comment:
We can't make it as a number of videos. It was my first idea, but Rust didn't allow me to do any operations with `N` to specify array size. That was a shame, that this didn't work:
```rust
struct Vertices<const N: usize>([Vertex; 4 * N]);
```
since:
```log
error: generic parameters may not be used in const operations
  --> compositor_render/src/renderer/common_pipeline/surface.rs:40:46
   |
40 | struct Vertices<const N: usize>([Vertex; 4 * N]);
   |                                              ^ cannot perform const operation using `N`
   |
   = help: const parameters may only be used as standalone arguments, i.e. `N`

```

I used `while` initially when this was a macro (for loops are not allowed in macros ¯\_(ツ)_/¯) and forgot to change it after copy-pasting :)

Reply:
Ohh, for loops are also forbidden in const functions :( 
I think we have to keep it as `while`

Reply:
sad

Reply:
all of this is sad but can't really do anything about it